yarpl/include/yarpl/Refcounted.h

```diff
@@ -66,9 +71,40 @@ class Reference {
   Reference() = default;
   inline /* implicit */ Reference(std::nullptr_t) {}
 
+  // TODO: Move check for count() into this constructor
+  // TODO: Remove public ctor for Reference(T*)
   explicit Reference(T* pointer) : pointer_(pointer) {
     inc();
   }
+  explicit Reference(T* pointer, detail::skip_initial_refcount_check)
+      : pointer_(pointer) {
+    inc();
+  }
+  explicit Reference(T* pointer, detail::do_initial_refcount_check)
+      : pointer_(pointer) {
+    /**
+     * consider the following:
+     *
+     class MyClass : Refcounted {
+       MyClass() {
+        // count() == 0
+        auto r = get_ref<MyClass>(this)
+        // count() == 1
+        do_something_with(r);
+        // if do_something_with(r) doens't keep a reference to r somewhere, then
+        // count() == 0
```

Comment:
Nit: This is a bit confusing, as this point `count() == 1` because `r` is still in scope.  I get what you mean though, `r` will go out of scope at the end of the ctor, decrement the count to 0, and then run the dtor.